### DISCUSSION TITLE
Strongswan High Availability setup documentation

### DISCUSSION THREAD

**Discussion by S733** [strongswan/strongswan#451] [Q&A]
Hi,
     I need help finding a proper documentation on setting up the High Availability feature using Strongswan ha plugin. The steps described on [wiki](https://wiki.strongswan.org/projects/strongswan/wiki/HighAvailability) are confusing for me. If I'm going to implement the HA feature between an on-premise pfsense to a cloud server (eg:AWS/GCP) how do I do it. Do I need three servers 1xIPsec gateway, 2xnodes. In the ha [test-scenarios](https://www.strongswan.org/testresults.html) page the servers carol and dave are connecting to private IP '192.168.0.5', is this IP added as a virtual IP on both nodes (alice and moon)?. How do we do the same config if we use a public IP.  Also there is a mention of a virtual gateway mars which I couldn't find in the architecture diagram. Any help would be much appreciated.

**Comment by tobiasbrunner** [comment]
> Do I need three servers 1xIPsec gateway, 2xnodes

No, two nodes both acting as VPN gateways.

> In the ha [test-scenarios](https://www.strongswan.org/testresults.html) page the servers carol and dave are connecting to private IP '192.168.0.5', is this IP added as a virtual IP on both nodes (alice and moon)?

Yes, they both have it installed and use it in the CLUSTERIP rules for the internal interfaces. Hosts behind the gateways can then route traffic to that IP, which gets sent via either gateway (depending on the assigned segments).

> How do we do the same config if we use a public IP

I guess the public IP would be the one configured in the CLUSTERIP rules for the public interfaces. Or what do you mean?

> Also there is a mention of a virtual gateway mars which I couldn't find in the architecture diagram.

The clients carol and dave connect to a "virtual" IP address (192.168.0.5) that both gateways use in their CLUSTERIP rules on the public interface. So they both together create a virtual gateway mars that's reachable on that IP.

**Reply by S733** [reply]
thanks @tobiasbrunner
> I guess the public IP would be the one configured in the CLUSTERIP rules for the public interfaces. Or what do you mean?

I don't know how we route the public IP to two nodes. We can add the IP as virtual IP inside the nodes, but if I'm accessing the public IP from the internet how will it reach the two nodes. Example:- if I add the public IP 104.xx.xx.xx as the virtual gateway and I configure roadwarriors carol and dave's ipsec.conf with right=104.xx.xx.xx, and I when I try to connect it can only reach  one node right. How do I reach the two nodes using single public IP.  Please correct me if I'm missing something, because of this confusion I asked in the OP if there is a gateway server before the nodes.


> iptables -A INPUT -i eth0 -d 192.168.0.200 -j CLUSTERIP --new \
>    --hashmode sourceip --clustermac 01:00:5e:00:00:20 \
>    --total-nodes 2 --local-node 0

In ubuntu 18.04 the above command returns the following error.
 iptables: Invalid argument. Run `dmesg' for more information.
 dmesg>  ipt_CLUSTERIP: bad local_nodes[0] 0

How do I clone the test-scenarios repo. I couldn't find the repo url anywhere

**Reply by tobiasbrunner** [reply]
> I don't know how we route the public IP to two nodes. We can add the IP as virtual IP inside the nodes, but if I'm accessing the public IP from the internet how will it reach the two nodes.

That's exactly what CLUSTERIP does. The nodes will respond with a virtual MAC address (`--clustermac`) to ARP requests for the shared IP address (`-d`) and both will receive the packets and decide based on the configuration (`--hashmode`) and assigned segments (strongSwan does that) whether to process it. You will need that or a similar mechanism (e.g. nftable's cluster match), otherwise it won't work.

The kernel patches for CLUSTERIP described on the [HA wiki page](https://wiki.strongswan.org/projects/strongswan/wiki/HighAvailability) do a even more and allow an active-active operation (without them, only active-passive will work).

If your clients supported IKEv2 redirection ([RFC 5685](https://datatracker.ietf.org/doc/html/rfc5685)), you could theoretically create a setup with a frontend gateway that directs clients to actual gateways (even more than two) without using virtual IPs. But you'd have a single point of failure in that frontend gateway.

> In ubuntu 18.04 the above command returns the following error.
> iptables: Invalid argument. Run `dmesg' for more information.
> dmesg> ipt_CLUSTERIP: bad local_nodes[0] 0

Yes, that won't work without our kernel patches. With 0, the node does not process any packets until strongSwan assigns a segment to it, the default kernel does not support that anymore, so just assign an actual value (1 or 2) if you use an unpatched kernel. 

> How do I clone the test-scenarios repo. I couldn't find the repo url anywhere

The test scenarios are all in the main repository ([`testing/tests/ha` directory](https://github.com/strongswan/strongswan/tree/master/testing/tests/ha) for the HA scenarios).
